trace/trace.go

```diff
@@ -317,43 +319,141 @@ func isTraceStateKeyValueValid(kv attribute.KeyValue) bool {
 		valueFormatRegExp.MatchString(kv.Value.Emit())
 }
 
-// SpanContext contains identifying trace information about a Span.
-type SpanContext struct {
+// SpanContextConfig contains mutable fields usable for constructing
+// an immutable SpanContext
+type SpanContextConfig struct {
 	TraceID    TraceID
 	SpanID     SpanID
 	TraceFlags byte
 	TraceState TraceState
 }
 
+// NewSpanContext constructs a SpanContext using values from the provided
+// SpanContextConfig
+func NewSpanContext(config SpanContextConfig) SpanContext {
+	return SpanContext{
+		traceID:    config.TraceID,
+		spanID:     config.SpanID,
+		traceFlags: config.TraceFlags,
+		traceState: config.TraceState,
+	}
+}
+
+// SpanContext contains identifying trace information about a Span.
+type SpanContext struct {
+	traceID    TraceID
+	spanID     SpanID
+	traceFlags byte
+	traceState TraceState
+}
+
 // IsValid returns if the SpanContext is valid. A valid span context has a
 // valid TraceID and SpanID.
 func (sc SpanContext) IsValid() bool {
 	return sc.HasTraceID() && sc.HasSpanID()
 }
 
+// TraceID returns the TraceID from the SpanContext
+func (sc SpanContext) TraceID() TraceID {
+	return sc.traceID
+}
+
 // HasTraceID checks if the SpanContext has a valid TraceID.
 func (sc SpanContext) HasTraceID() bool {
-	return sc.TraceID.IsValid()
+	return sc.traceID.IsValid()
+}
+
+// WithTraceID returns a new SpanContext with the TraceID replaced.
+func (sc SpanContext) WithTraceID(traceID TraceID) SpanContext {
+	return SpanContext{
+		traceID:    traceID,
+		spanID:     sc.spanID,
+		traceFlags: sc.traceFlags,
+		traceState: sc.traceState,
+	}
+}
+
+// SpanID returns the SpanID from the SpanContext
+func (sc SpanContext) SpanID() SpanID {
+	return sc.spanID
 }
 
 // HasSpanID checks if the SpanContext has a valid SpanID.
 func (sc SpanContext) HasSpanID() bool {
-	return sc.SpanID.IsValid()
+	return sc.spanID.IsValid()
+}
+
+// WithSpanID returns a new SpanContext with the SpanID replaced.
+func (sc SpanContext) WithSpanID(spanID SpanID) SpanContext {
+	return SpanContext{
+		traceID:    sc.traceID,
+		spanID:     spanID,
+		traceFlags: sc.traceFlags,
+		traceState: sc.traceState,
+	}
+}
+
+// TraceFlags returns the flags from the SpanContext
+func (sc SpanContext) TraceFlags() byte {
+	return sc.traceFlags
+}
+
+// WithTraceFlags returns a new SpanContext with the TraceFlags replaced.
+func (sc SpanContext) WithTraceFlags(flags byte) SpanContext {
```

Comment:
I think that the `With` func are nice but not required now, I would limit the public API initially and add utilities based on future feedback.

Reply:
There were a number of existing uses where a `SpanContext` was derived from an existing `SpanContext`, typically from a parent.  These make that pattern simpler and less error prone compared to reconstructing from exported accessors at the consuming end.  It also makes it so that any future additions to the `SpanContext` struct can be accounted for here in these methods without worrying that a consumer may have constructed incomplete derived contexts.